### ISSUE TITLE
Shouldn't Complex#pow() handle Complex.ZERO as special case?

### ISSUE THREAD

**Issue by axkr** [Hipparchus-Math/hipparchus#74]
Shouldn't `Complex#pow()` handle Complex.ZERO as special case?
- https://github.com/Hipparchus-Math/hipparchus/blob/16c433c7804b4998cbf31895bee9e313544356bf/hipparchus-core/src/main/java/org/hipparchus/complex/Complex.java#L894

```
@Test
    public void testPowDouble() {
        Complex x = Complex.ZERO;
        Complex expected = Complex.ZERO;
        UnitTestUtils.assertEquals(expected, x.pow(2.0), 1.0e-5);
    }
```

**Comment by axkr** [comment]
See:
-     https://github.com/apache/commons-numbers/blob/817cb6d4c8ec6afdfed94e9d5e8bbd50e0b9f9db/commons-numbers-complex/src/main/java/org/apache/commons/numbers/complex/Complex.java#L2621
- https://github.com/apache/commons-numbers/blob/817cb6d4c8ec6afdfed94e9d5e8bbd50e0b9f9db/commons-numbers-complex/src/main/java/org/apache/commons/numbers/complex/Complex.java#L2653

**Comment by maisonobe** [comment]
Yes, zero should be handled specifically. In fact, many numbers should be handled specifically, as the double version of pow has as much as 17 special cases! The weirdest one being that x^0 should be 1.0 regardless of x, even if x = NaN (this is the only operation I known of that returns regular numbers when applied to a NaN).

So this is definitely a bug.

**Comment by psteitz** [comment]
Thanks for the feedback.  We have stayed with the simple, NaN-generating 
formulas for undefined values such as below in Hipparchus.  The NaN 
value returned there is consistent with what is specified by the 
javadoc, which defines pow using the log function, making it undefined 
at ZERO.

**Comment by psteitz** [comment]
Sorry, I did not see this before responding.  Why, exactly and what is 
rationale for extending the definition?

Phil

**Comment by axkr** [comment]
@psteitz `Complex.ZERO.pow(<int-number>)` is a common case you can't prevent in more complicated formulas. The commons-numbers seem to implement these cases.

**Comment by psteitz** [comment]
Please describe the actual use cases and rationale for the definition 
that you would like to make.  Returning ad hoc results that allow 
computations to proceed does not make sense unless there is some kind of 
mathematical definition that allows you to meaningfully extend the 
analytical definition.  What is that definition?  Can you provide 
references that explain why this choice is 0) meaningful 1) best for the 
widest range of cases 2) worth the performance and code maintenance hit?

**Comment by maisonobe** [comment]
I think at least Complex.pow should be compatible with {Math, StrictMath, FastMath}.pow when both numbers are real, as Complex is an extension of the Real field.

As mentioned in the language official documentation for [Math.pow](https://docs.oracle.com/en/java/javase/13/docs/api/java.base/java/lang/Math.html#pow(double,double)) and [StrictMath.pow](https://docs.oracle.com/en/java/javase/13/docs/api/java.base/java/lang/StrictMath.html#pow(double,double)), there are a number of special cases for the function and the order in which they are tested is important. For example the fact the second argument is tested for zero (and 1.0 is returned in that case, even if first argument is NaN) is tested before the others and in addition some provisions are made to avoid one case being addressed by different parts of the if, so for example 0^0 us handled by the first test and returns 1.0, and not by the eights test which would return either positive 0 or positive infinity. Note also that the commons number implementation is IMHO wrong here, 0^0 should be 1.0, not 0 and 0^y when y is not zero should never be NaN.

I suggest to do the following:

```
public Complex pow(Complex y) {
  if (y.imaginary == 0.0) {
    if (this.imaginary == 0.0) {
       // use real implementation that handles a bunch of special cases
       return createComplex(FastMath.pow(this.real, y.real), 0.0);
    } else if (y.real == 0.0) {
       // this^0, when this is not a real number
       return ONE;
    }
  } else {
    // generic implementation
    return  this.log().multiply(x).exp();
  }
}
```

**Comment by maisonobe** [comment]
The issue is fixed. The solution is not the one I exposed in a previous comment, because it was wrong: typically (-1 + 0i)^½ would have been none instead of ±i.
Working on the fix, it appeared to me that other cases were wrong/not logical. For example (∞ + i)^1 should either be (∞ + i) or Complex.INF but instead it was Complex.NaN. There were other similar cases.
Last thing I saw was that integer powers were computed with complex log and exponential when more accurate and stable fast computation (based on successive squarings) could be used, so I also added a new signature for power integers. Indeed, this new signature will be needed for issue #67. This new method Complex.pow(int) should be used when possible, but real powers with integer values are detected even when the argument is passed as a Complex instance.
